### DISCUSSION TITLE
HPIPM in the MPC QP benchmark?

### DISCUSSION THREAD

**Discussion by rienq** [qpsolvers/mpc_qpbenchmark#7] [General]
Hi,

First of all, nice work on this MPC benchmarking test framework for QP solvers!
I noticed that HPIPM is mentioned in the list of QP solvers, but I do not see it in the benchmark results, why is that?

Thank you,
Rien

**Comment by stephane-caron** [comment]
Thank you for pointing this out :+1:

HPIPM was likely not installed on the machine I last run the benchmark on. (While most solvers can be installed straight from the conda ``environment.yaml`` file, Python bindings for HPIPM are [installed from source](https://github.com/giaf/hpipm?tab=readme-ov-file#python).) I will make sure it is added next time I run the benchmark.

In the meantime, you can also [run](https://github.com/qpsolvers/mpc_qpbenchmark?tab=readme-ov-file#installation) the benchmark on your machine, and feel free to report any issue you may encounter.

**Reply by stephane-caron** [reply]
I've re-run the test set with HPIPM on the same machine and reported the results in https://github.com/qpsolvers/mpc_qpbenchmark/discussions/8. Hoping this helps!

**Comment by giaf** [comment]
Hi,

first of all, thanks for adding also HPIPM to the solvers and benchmarking it.
Looking into the results, I found some issues that I would like to report:

1) Strangely HPIPM success rate is lower at low accuracy, and it increases at mid and even more at high accuracy.
This behavior doesn't appear for the other solvers.
The reason seems to be the low percentage of problems where "solved" return codes are correct.
Looking into the issue more carefully, it seems that the issue is related to the duality gap: the interfaces sets the solver 'max_res_comp' (i.e. the complementarity slackness) to a specified value, but then it accepts the solution as correct if the duality gap (instead of the complementarity slackness) is within that specified value.
Complementarity slackness and duality gap are equivalent only for a feasible primal and dual point, but numerically this is true only approximately, and the approximation can be especially crude for the low accuracy case, where also primal and dual feasibility tolerances are set to low values.

Currently HPIPM does not support a termination criterion based on duality gap, but this could be easily added if requested.
In any case, as it stands I would say that the interface is "wrong" as it asks for something and checks something else: at least there should be a note explaining the reason for the unusually low success rate of HPIPM.

2) Another issue is common to all solvers that cannot exploit generic sparsity in the problem.
Looking at the interfaces source code, it seems that in case of solvers that _can_ exploit generic sparsity in the problem such as OSQP, clarabel, qpalm, scs for example, the matrix data is passed as an union of ndarray and csc_matrix, and thus the solvers can exploit the sparsity in the problem formulation.
On the other hand, , it seems that in case of solvers that can _not_ exploit generic sparsity in the problem such as quadprog, HPIPM, qpOASES or DAQP for example, all problem data is passed only as ndarray, and there seem to be no pre-processing steps such as condensing: this means that the sparse multi-stage MPC problem is cast as a dense one with a bunch of structural zeros too.
While this could be understandable in benchmarking these solvers on generic QPs, in the special case of MPC QP benchmarking this leads to at least misleading conclusions, as this is not the way such MPC-tailored solvers are designed and expected to be used.
Dense QP solvers are always coupled with condensing routines when solving MPC problems, such that the dynamics equality constraints are exploited to remove the state variable form the problem formulation, leaving only the true degrees of freedom (i.e. the control variables) as the optimization variables.
The OCP QP solver in HPIPM can also directly handle the multistage structure of the MPC problems, such that the fixed pattern of the structural zeros is exploited.
Clearly this requires to explicitly expose this structure to the solver as it is not automatically detected; however this is how HPIPM is generally used in the MPC community.

I understand that adding preprocessing steps such as condensing or a structure detection is outside the scope of this benchmark and would also raise the question of picking the best formulation for each solver for each specific problem.
But at least there should be a big blinking warning ( ;-) ) telling what is going on in the interfaces, and that this is not the way these solvers are expected to be used in solving MPC problems and that most likely much better performance can be obtained by reformulating the problem in the way these solvers are designed for.

There is already a "Known limitations" section that already contains factors affecting performance to a much smaller degree such as thermal throttling, so the issues above could definitely find space there too.

Thank you,

Gianluca

**Reply by stephane-caron** [reply]
Thank you for this detailed feedback!

## 1. Complementarity slackness vs duality gap

This is indeed an issue, solvers should be asked the same tolerance "contract" and be evaluated against the same criteria. This point is not specific to the MPC qpbenchmark and holds for all [qpbenchmark](https://github.com/qpsolvers/qpbenchmark) test sets, therefore I suggest following up in the following issue: https://github.com/qpsolvers/qpbenchmark/issues/122

>  In any case, as it stands I would say that the interface is "wrong" as it asks for something and checks something else: at least there should be a note explaining the reason for the unusually low success rate of HPIPM.

I've added the open issue to the *Known limitations* section upstream https://github.com/qpsolvers/qpbenchmark/commit/5c8a84b1fce915801d951169adcfdd0dfdee85f4.

## 2. Sparsity and condensing

As the first part of your point is general to all solvers with a "dense API", I've spun it off to https://github.com/qpsolvers/mpc_qpbenchmark/discussions/9 to give it more visibility and organize the discussion.

> The OCP QP solver in HPIPM can also directly handle the multistage structure of the MPC problems, such that the fixed pattern of the structural zeros is exploited.
> Clearly this requires to explicitly expose this structure to the solver as it is not automatically detected; however this is how HPIPM is generally used in the MPC community.

Do you expect there would be an improvement when problems (such as the ones in the test set so far) are already condensed, and there is more of an "upper block triangular" rather than "block diagonal" structure?

So far we haven't evaluated cases where additional problem structure is given to some solvers in the benchmark (we just conform to their API when they expect sparse or dense matrices). One way to keep it fair and informative could be to have run the solver with/without the extra info.

> But at least there should be a big blinking warning ( ;-) ) telling what is going on in the interfaces, and that this is not the way these solvers are expected to be used in solving MPC problems and that most likely much better performance can be obtained by reformulating the problem in the way these solvers are designed for.

I fully agree. I've reproduced the *Known limitations* section in this test set's readme along with a warning in https://github.com/qpsolvers/mpc_qpbenchmark/commit/f767fa39942e4d68ce27a2a903e093798b29dc7c

Another suggestion: if you have documentation on best practices to use HPIPM on MPC problems, don't hesitate to open a PR adding a link to it in the *See also* section of the readme here.
